lib/course_planner/tasks.ex

```diff
@@ -47,59 +23,81 @@ defmodule CoursePlanner.Tasks do
-    |> where([t], is_nil(t.user_id))
     |> where([t], t.finish_time > ^now)
     |> Repo.all()
+    |> Enum.reject(fn(task) ->
+           length(task.volunteers) >= task.max_volunteers or
+           Enum.any?(task.volunteers, &(&1.id == id))
+       end)
   end
 
   def get_past(sort_opt, id, now) do
     sort_opt
     |> task_query()
-    |> where([t], t.user_id == ^id)
     |> where([t], t.finish_time < ^now)
     |> Repo.all()
+    |> Enum.reject(fn(task) ->
+         not Enum.any?(task.volunteers, &(&1.id == id))
+       end)
   end
 
   def get_for_user(sort_opt, id, now) do
     sort_opt
     |> task_query()
-    |> where([t], t.user_id == ^id)
     |> where([t], t.finish_time > ^now)
     |> Repo.all()
+    |> Enum.reject(fn(task) ->
+         not Enum.any?(task.volunteers, &(&1.id == id))
+       end)
   end
 
-  def task_query(sort_opt), do: Task |> sort(sort_opt) |> preload(:user)
+  def task_query(sort_opt) do
+     Task
+     |> sort(sort_opt)
+     |> join(:left, [t], v in assoc(t, :volunteers))
+     |> preload([t, v], [volunteers: v])
+  end
 
   defp sort(query, nil), do: query
   defp sort(query, "fresh"), do: order_by(query, [t], desc: t.updated_at)
   defp sort(query, "closest"), do: order_by(query, [t], asc: t.finish_time)
 
-  def grab(task_id, user_id, now) do
-    with {:ok, task}              <- get(task_id),
-      %{valid?: true} = changeset <- Task.changeset(task),
-      {:ok, changeset}            <- validate_finish_time(changeset, now),
-      {:ok, changeset}            <- validate_already_assigned(changeset, task)
-    do
+  def grab(task_id, volunteer_id) do
+    with {:ok, task} <- get(task_id),
+        %{valid?: true} = changeset <- Task.changeset(task)
+     do
+      new_volunteer = Volunteers.get!(volunteer_id)
+      updated_volunteer_list = [new_volunteer | task.volunteers]
+
       changeset
-      |> Changeset.put_change(:user_id, user_id)
+      |> Task.update_volunteer(:volunteers, updated_volunteer_list)
       |> Repo.update()
     else
       error -> error
     end
   end
 
-  defp validate_finish_time(changeset, now) do
-    fin = Changeset.get_field(changeset, :finish_time)
-    case Timex.compare(fin, now) do
-      1  -> {:ok, changeset}
-      -1 -> {:error, :already_finished}
-      _  -> {:error, :unkown}
+  def drop(task_id, volunteer_id) do
+    with {:ok, task} <- get(task_id),
+        %{valid?: true} = changeset <- Task.changeset(task)
+     do
+      drop_volunteer = Volunteers.get!(volunteer_id)
+      updated_volunteer_list = List.delete(task.volunteers, drop_volunteer)
+
+      changeset
+      |> Task.drop_volunteer(:volunteers, updated_volunteer_list)
+      |> Repo.update()
+    else
+      error -> error
     end
   end
 
-  defp validate_already_assigned(changeset, %{user_id: nil}), do: {:ok, changeset}
-  defp validate_already_assigned(_changeset, _), do: {:error, :already_assigned}
+  def update_changeset_volunteers(changeset, task_params, volunteer_param_name) do
```

Comment:
then it is probably more interesting to receive volunteer_ids straight from the controller instead of params and the param name, because this way you still depend on the params here